go/parquet/internal/encoding/byte_stream_split.go

```diff
@@ -0,0 +1,367 @@
+	encodeByteStreamSplitWidth4(enc.flushBuffer.Bytes(), in.Bytes())
+	return enc.flushBuffer.Finish(), nil
+}
+
+func (enc *ByteStreamSplitFloat32Encoder) Release() {
+	enc.PlainFloat32Encoder.Release()
+	releaseBufferToPool(enc.flushBuffer)
+	enc.flushBuffer = nil
+}
+
+// ByteStreamSplitFloat64Encoder writes the underlying bytes of the Float64
+// into interlaced streams as defined by the BYTE_STREAM_SPLIT encoding
+type ByteStreamSplitFloat64Encoder struct {
+	PlainFloat64Encoder
+	flushBuffer *PooledBufferWriter
+}
+
+func (enc *ByteStreamSplitFloat64Encoder) FlushValues() (Buffer, error) {
+	in, err := enc.PlainFloat64Encoder.FlushValues()
+	if err != nil {
+		return nil, err
+	}
+
+	if enc.flushBuffer == nil {
+		enc.flushBuffer = NewPooledBufferWriter(in.Len())
+	}
+
+	enc.flushBuffer.buf.Resize(in.Len())
+	encodeByteStreamSplitWidth8(enc.flushBuffer.Bytes(), in.Bytes())
+	return enc.flushBuffer.Finish(), nil
+}
+
+func (enc *ByteStreamSplitFloat64Encoder) Release() {
+	enc.PlainFloat64Encoder.Release()
+	releaseBufferToPool(enc.flushBuffer)
+	enc.flushBuffer = nil
+}
+
+// ByteStreamSplitInt32Encoder writes the underlying bytes of the Int32
+// into interlaced streams as defined by the BYTE_STREAM_SPLIT encoding
+type ByteStreamSplitInt32Encoder struct {
+	PlainInt32Encoder
+	flushBuffer *PooledBufferWriter
+}
+
+func (enc *ByteStreamSplitInt32Encoder) FlushValues() (Buffer, error) {
+	in, err := enc.PlainInt32Encoder.FlushValues()
+	if err != nil {
+		return nil, err
+	}
+
+	if enc.flushBuffer == nil {
+		enc.flushBuffer = NewPooledBufferWriter(in.Len())
+	}
+
+	enc.flushBuffer.buf.Resize(in.Len())
+	encodeByteStreamSplitWidth4(enc.flushBuffer.Bytes(), in.Bytes())
+	return enc.flushBuffer.Finish(), nil
+}
+
+func (enc *ByteStreamSplitInt32Encoder) Release() {
+	enc.PlainInt32Encoder.Release()
+	releaseBufferToPool(enc.flushBuffer)
+	enc.flushBuffer = nil
+}
+
+// ByteStreamSplitInt64Encoder writes the underlying bytes of the Int64
+// into interlaced streams as defined by the BYTE_STREAM_SPLIT encoding
+type ByteStreamSplitInt64Encoder struct {
+	PlainInt64Encoder
+	flushBuffer *PooledBufferWriter
+}
+
+func (enc *ByteStreamSplitInt64Encoder) FlushValues() (Buffer, error) {
+	in, err := enc.PlainInt64Encoder.FlushValues()
+	if err != nil {
+		return nil, err
+	}
+
+	if enc.flushBuffer == nil {
+		enc.flushBuffer = NewPooledBufferWriter(in.Len())
+	}
+
+	enc.flushBuffer.buf.Resize(in.Len())
+	encodeByteStreamSplitWidth8(enc.flushBuffer.Bytes(), in.Bytes())
+	return enc.flushBuffer.Finish(), nil
+}
+
+func (enc *ByteStreamSplitInt64Encoder) Release() {
+	enc.PlainInt64Encoder.Release()
+	releaseBufferToPool(enc.flushBuffer)
+	enc.flushBuffer = nil
+}
+
+type ByteStreamSplitFloat32Decoder = ByteStreamSplitDecoder[float32]
+type ByteStreamSplitFloat64Decoder = ByteStreamSplitDecoder[float64]
+type ByteStreamSplitInt32Decoder = ByteStreamSplitDecoder[int32]
+type ByteStreamSplitInt64Decoder = ByteStreamSplitDecoder[int64]
```

Comment:
Should we do the same approach for the encoders too? 

Should probably also add godoc comments on these

Reply:
Just added the godoc comments.

I did like how the generic decoders came out and looked at what it would take to do the same for encoders. It's a little tricker with the encoders because they all embed their respective "Plain" encoders. It's awkward to make this generic at the moment because the Plain encoders are not generic themselves. I think this all gets a lot simpler if/when the overall refactor of parquet to use generics is done, since then the `ByteStreamSplitEncoder[T]` could just embed `PlainEncoder[T]` once it exists.